ASoC: SOF: ipc4: Add support for split firmware releases

ujfalusi opened the pull request:
A split SOF release consists of a base firmware and two libraries:
<fw_filename>-openmodules.ri for processing (audio) modules
<fw_filename>-debug.ri for debug and developer modules

To handle this new release model add infrastructure to try to load the two
library after boot optionally.

This approach will allow flexibility on handling platforms in sof-bin with
single or split configuration:
single release: base firmware only
split release: base firmware + openmodules + debug

The files for the split firmware are located at the firmware directory.

The documentation update is: https://github.com/thesofproject/sof-docs/pull/505

ujfalusi committed:
ASoC: SOF: ipc4: Add support for split firmware releases

A split SOF release consists of a base firmware and two libraries:
<fw_filename>-openmodules.ri for processing (audio) modules
<fw_filename>-debug.ri for debug and developer modules

To handle this new release model add infrastructure to try to load the two
library after boot optionally.

This approach will allow flexibility on handling platforms in sof-bin with
single or split configuration:
single release: base firmware only
split release: base firmware + openmodules + debug

The files for the split firmware are located at the firmware directory.

Signed-off-by: Peter Ujfalusi <peter.ujfalusi@linux.intel.com>

lyakh commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -256,26 +262,80 @@ static int sof_ipc4_load_library_by_uuid(struct snd_sof_dev *sdev,
+
+		kfree(lib_filename);
+		if (ret)
+			break;
```
why `break`? If any of the libraries fail loading you don't attempt the rest? In this case it kind of makes sense - if there's no "main" module bundle, there's probably no "debug" either, but isn't the general concept - here's a list of optional libraries, we try to load whatever is available?

ujfalusi replied:
The `sof_ipc4_load_library()` will only return error if there were real error for the loading, the bundle not present is not an error, but if the bundle loading fails then that is an error, something is not right and we will abort the probing altogethe.

lgirdwood replied:
We should continue if a bundle fails, this way we still get base FW (and it may have some built-in modules) and a registered sound card. Yes we tell the user that we could not load the library and audio may not fully work as expected.

ujfalusi replied:
We will fail to probe the card as the modules form the failed library are missing and the topology will not probe.

ujfalusi replied:
OK, if we continue then the car will not come up, but the debugfs for SOF is going the be there to provide further clues.

ujfalusi commented:
Changes since v1:
- follow the sof-docs update [1] and look for the base libraries in the same directory as the firmware

[1] https://github.com/thesofproject/sof-docs/pull/506

ujfalusi commented:
Changes since v2:
- Documentation as inline comment for the sof_ipc4_load_library_bundles() function
- fix a memory leak (after kasprintf) and added missing NULL check after kzalloc

ujfalusi commented:
Changes since v3:
- Print warning on library loading error. The topology loading will likely fail and no card will be present, but the SOF debugfs is going to be available.

ranj063 commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -202,21 +195,28 @@ static int sof_ipc4_load_library_by_uuid(struct snd_sof_dev *sdev,
+		ret = firmware_request_nowarn(&fw_lib->sof_fw.fw, lib_filename,
+					      sdev->dev);
+		if (ret < 0) {
+			dev_dbg(sdev->dev, "Library file '%s' is not present\n",
```
@ujfalusi with the current releases where the openmodule/debug binaries are absent, will we always see this debug message? I think this will too much of a distraction. How about we check the presence of these files before requesting to load the libraries?

ujfalusi replied:
Or not print here at all if does not exist...

ujfalusi commented:
Changes since v4:
- Drop the debug print if the optional library is not found in filesystem
- Change terms to single and split firmware release (two types of the modular firmware).

ranj063 commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -256,26 +254,111 @@ static int sof_ipc4_load_library_by_uuid(struct snd_sof_dev *sdev,
+ * sof_ipc4_load_library_bundles - loads the library parts of a modular firmware
+ * @sdev: SOF device
+ *
+ * With IPC4 the firmware can be monolithic or modular release.
```
@ujfalusi you want to remove the monolithic reference in this PR too?

ujfalusi replied:
yes, I wanted to... ;)

ujfalusi commented:
Changes since v5:
- replace references to modular to split firmware
- change `sof_ipc4_load_library_bundles()`  to `sof_ipc4_complete_split_release()` to better align with the functionality

bardliao commented:
SOFCI TEST

bardliao commented:
SOFCI TEST

lyakh reviewed (approved):
works with https://github.com/thesofproject/sof/pull/9702

ujfalusi commented:
@ranj063, can you check the patch again?

lyakh commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -256,26 +254,111 @@ static int sof_ipc4_load_library_by_uuid(struct snd_sof_dev *sdev,
+	if (!lib_name_base)
+		return -ENOMEM;
+
+	strscpy(lib_name_base, fw_filename, sizeof(lib_name_base));
```
can you check the size please?

ujfalusi replied:
which size?

lyakh replied:
there's only one size - the last argument of `strscpy()`. It's a size of a pointer, is that what is needed here?

ujfalusi replied:
valid, it did worked on most platform with sof-123.ri pattern ;)

lyakh replied:
it's because on 64-bit systems `sizeof(char *) == 8`

lyakh commented on sound/soc/sof/ipc4-loader.c:
```diff
@@ -256,26 +254,111 @@ static int sof_ipc4_load_library_by_uuid(struct snd_sof_dev *sdev,
 
-	kfree(fw_filename);
-
+	(*lib_id)++;
```
what sense does it make incrementing this if the caller then ignores it?

ujfalusi replied:
it is in the for loop, the lib_id is incremented for the next call?

ujfalusi replied:
I think there were a reason why I did this, but cannot recall, I guess the increment can be done in a caller function.
Likely I had a different implementation which I reworked to arrive to this one..

lyakh replied:
ah, right, it's a loop in `sof_ipc4_complete_split_release()`, I only looked at `sof_ipc4_load_library_by_uuid()` where it's just a single call. Ok, it's a bit confusing there, that a value is incremented but then ignored, but I see that it's actually ok. If you make a change, maybe just increment the `lib_id` in the loop instead of in the function, then you also don't need to change an `int` to a pointer at multiple locations in the function body

lyakh reviewed (changes requested):
2 comments seem too suspicious

ujfalusi commented:
Changes since v6:
- Correct strscpy() size parameter and add comments to explain what is going on
- There is no real need to pass the lib_id as pointer to sof_ipc4_load_library(), it can be incremented in sof_ipc4_complete_split_release() if needed.

ujfalusi commented:
@lyakh, thank you for pointing the issues out, the strscpy() was a really bad from me :(
